### PR TITLE
(#600) Move Report Sender from Installation Manager server to WS master server

### PR DESCRIPTION
**Pull request by dmytro-ndp**
@sleshchenko, @skabashnyuk, @AndrienkoAleksandr: please, review this request.

Signed-off-by: Dmytro Nochevnov <dnochevnov@codenvy.com>

### CODE REVIEW

**Comment by AndrienkoAleksandr** [comment]
Ok, for me.

File: wsmaster/codenvy-hosted-report/pom.xml [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+            <artifactId>guice</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>javax.inject</groupId>
+            <artifactId>javax.inject</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>javax.mail</groupId>
+            <artifactId>mail</artifactId>
+            <version>1.4.4</version>
```
Comment (skabashnyuk) [inline wsmaster/codenvy-hosted-report/pom.xml]: version

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java [hunk 2]
```diff
@@ -47,18 +51,29 @@ public ReportService() {
     @GET
     @Path("/parameters/{report_type}")
```
Comment (skabashnyuk) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: What kind of authorisation is used for this class? Who can call methods of this class?

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/UtilService.java [hunk 3]
```diff
@@ -41,15 +43,17 @@
     public UtilService() {
     }
 
-    /** Get client's IP. */
+    /**
+     * Get client's IP.
```
Comment (skabashnyuk) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/UtilService.java]: Why do we need this method?

File: wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/LicenseService.java [hunk 4]
```diff
@@ -157,10 +142,10 @@ public Response getLicenseProperties() throws ApiException {
...
 
             Map<String, String> properties = features
-                    .entrySet()
-                    .stream()
-                    .map(entry -> new AbstractMap.SimpleEntry<>(entry.getKey().toString(), entry.getValue()))
-                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+                .entrySet()
+                .stream()
+                .map(entry -> new AbstractMap.SimpleEntry<>(entry.getKey().toString(), entry.getValue()))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
```
Comment (skabashnyuk) [inline wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/LicenseService.java]: may be http://stackoverflow.com/a/22743672/318536?

File: wsmaster/codenvy-hosted-report-shared/pom.xml [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    <packaging>jar</packaging>
+    <name>Codenvy :: Hosted :: Report :: Shared</name>
+    <properties>
+        <dto-generator-out-directory>${project.build.directory}/generated-sources/dto/</dto-generator-out-directory>
+    </properties>
+    <dependencies>
+        <dependency>
+            <groupId>com.google.code.gson</groupId>
+            <artifactId>gson</artifactId>
+            <version>2.3.1</version>
```
Comment (skabashnyuk) [inline wsmaster/codenvy-hosted-report-shared/pom.xml]: version

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java (hunk 2 above)
Comment (dmytro-ndp) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: There is no authorization there. Anyone could call that method.

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/UtilService.java (hunk 3 above)
Comment (dmytro-ndp) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/UtilService.java]: To identify the user who is using Codenvy.

File: wsmaster/codenvy-hosted-report/pom.xml (hunk 1 above)
Comment (dmytro-ndp) [inline wsmaster/codenvy-hosted-report/pom.xml]: okay

File: wsmaster/codenvy-hosted-report-shared/pom.xml (hunk 5 above)
Comment (dmytro-ndp) [inline wsmaster/codenvy-hosted-report-shared/pom.xml]: okay

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java [hunk 6]
```diff
@@ -47,18 +51,29 @@ public ReportService() {
...
     @Produces(MediaType.APPLICATION_JSON)
-    public Response getReportParameters(@PathParam("report_type") String reportType) {
+    public Response getReportParameters(@PathParam("report_type") String reportTypeStr) {
         try {
-            if (reportType == null || ReportType.valueOf(reportType.toUpperCase()) == null) {
-                throw new RuntimeException("Report type is unknown.");
-            }
+            Objects.requireNonNull(reportTypeStr, "Report type is unknown.");
```
Comment (sleshchenko) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: @PathParam can't equal to null

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java [hunk 7]
```diff
@@ -47,18 +51,29 @@ public ReportService() {
...
 
-            ReportParameters parameters = ReportType.valueOf(reportType.toUpperCase()).getParameters();
+            ReportType reportType = ReportType.valueOf(reportTypeStr.toUpperCase());
+            Objects.requireNonNull(reportType, "Report type is unknown.");
```
Comment (sleshchenko) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: Are you sure that you should throw NPE here which will be mapped to 500 Internal Server Error. Mb 401 Bad request or 404 Not found would be better here

File: wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/LicenseService.java (hunk 4 above)
Comment (dmytro-ndp) [inline wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/LicenseService.java]: The code is okay as for me. Please, explain your question a bit more.

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java (hunk 7 above)
Comment (dmytro-ndp) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: 404 Not Found is more suitable her. Thank you!

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java (hunk 6 above)
Comment (dmytro-ndp) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: You are right. Thank you for note.

File: wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/LicenseService.java (hunk 4 above)
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/LicenseService.java]: You can rewrite this as
```
Map<String, String> properties = features.entrySet()
                                                     .stream()
                                                     .collect(Collectors.toMap(entry -> entry.getKey().toString(),
                                                                               Map.Entry::getValue));
```

File: wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/LicenseService.java [hunk 8]
```diff
@@ -183,61 +168,34 @@ public Response getLicenseProperties() throws ApiException {
...
+            throw new ServerException("Failed to check if Codenvy usage matches Codenvy License constraints. ", e);
         }
     }
 
     @GET
     @Path("/legality/node")
     @Produces(MediaType.APPLICATION_JSON)
     @ApiOperation(value = "Check if Codenvy node usage matches Codenvy License constraints, or free usage rules.",
-                  notes = "If nodeNumber parameter is absent then actual number of machine nodes will be validated")
+        notes = "If nodeNumber parameter is absent then actual number of machine nodes will be validated")
```
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/LicenseService.java]: please revert these changes. New formatting is wrong

File: wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java [hunk 9]
```diff
@@ -40,23 +43,31 @@
...
+        this.adminUserDao = adminUserDao;
+        this.dockerConnector = dockerConnector;
     }
 
     /**
      * Stores valid Codenvy license into the storage.
      *
      * @throws NullPointerException
-     *         if {@code codenvyLicense} is null
+     *         if {@code licenseStr} is null
```
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java]: I prefer to no use variable type in name of it. Mb licenseText or licenseContent or just license would be better?

File: wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java [hunk 10]
```diff
@@ -108,4 +119,37 @@ public void delete() throws LicenseException {
             throw new LicenseException(e.getMessage(), e);
         }
     }
+
+    /**
+     * Return true if only Codenvy usage meets the constrains of license properties or free usage properties.
+     **/
+    public boolean isCodenvyUsageLegal() throws ServerException, IOException {
+        long actualUsers = adminUserDao.getAll(30, 0).getTotalItemsCount();   // TODO ndp (codenvy/600) fix it get all users number
```
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java]: Can you explain meaning and benefit of `ndp (codenvy/600)`

File: wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java [hunk 11]
```diff
@@ -108,4 +119,37 @@ public void delete() throws LicenseException {
...
+            return codenvyLicense.isLicenseUsageLegal(actualUsers, actualServers);
+        } catch (LicenseException e) {
+            return CodenvyLicense.isFreeUsageLegal(actualUsers, actualServers);
+        }
+    }
+
+    /**
+     * Return true if only node number meets the constrains of license properties or free usage properties.
+     * If nodeNumber == null, uses actual number of machine nodes.
+     * @param nodeNumber - number of machine nodes.
```
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java]: Fix formatting here

File: wsmaster/codenvy-hosted-report-shared/pom.xml (hunk 5 above)
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-report-shared/pom.xml]: You should not use version here

File: wsmaster/codenvy-hosted-report-shared/pom.xml [hunk 12]
```diff
@@ -0,0 +1,163 @@
...
+                        <version>${project.version}</version>
+                    </dependency>
+                </dependencies>
+                <configuration>
+                    <dtoPackages>
+                        <package>com.codenvy.report.shared.dto</package>
+                    </dtoPackages>
+                    <outputDirectory>${dto-generator-out-directory}</outputDirectory>
+                    <genClassName>com.codenvy.report.shared.dto.DtoServerImpls</genClassName>
+                    <impl>server</impl>
```
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-report-shared/pom.xml]: Ideally you should not generate dto impls for server code in shared module

File: wsmaster/codenvy-hosted-report-shared/src/main/java/com/codenvy/report/shared/dto/Ip.java [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+ */
+package com.codenvy.report.shared.dto;
+
+import org.eclipse.che.dto.shared.DTO;
+
+/**
+ * @author Dmytro Nochevnov
+ */
+@DTO
+public interface Ip {
```
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-report-shared/src/main/java/com/codenvy/report/shared/dto/Ip.java]: Mb it is better to name this class as IpDto? As far as I know It is common way for us to add Dto suffix to dto interfaces

File: wsmaster/codenvy-hosted-report/pom.xml (hunk 1 above)
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-report/pom.xml]: remove version here

File: wsmaster/codenvy-hosted-report/pom.xml [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+            <resource>
+                <directory>src/main/java</directory>
+            </resource>
+            <resource>
+                <directory>src/main/resources</directory>
+            </resource>
+        </resources>
+        <plugins>
+            <plugin>
+                <artifactId>maven-compiler-plugin</artifactId>
```
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-report/pom.xml]: Are you sure that you need this section?

File: wsmaster/codenvy-hosted-report/src/main/java/com/codenvy/report/ReportModule.java [hunk 15]
```diff
@@ -0,0 +1,32 @@
...
+/**
+ * @author Dmytro Nochevnov
+ */
+@DynaModule
+public class ReportModule extends AbstractModule {
+    @Override
+    protected void configure() {
+        bind(ReportSender.class);
+
+        install(new ScheduleModule());
```
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-report/src/main/java/com/codenvy/report/ReportModule.java]: Will this module be deployed to api.war? It yes then you can remove this installing.

File: wsmaster/codenvy-hosted-report/src/main/java/com/codenvy/report/ReportModule.java [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+package com.codenvy.report;
+
+import com.google.inject.AbstractModule;
+import org.eclipse.che.commons.schedule.executor.ScheduleModule;
+import org.eclipse.che.inject.DynaModule;
+
+/**
+ * @author Dmytro Nochevnov
+ */
+@DynaModule
```
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-report/src/main/java/com/codenvy/report/ReportModule.java]: It would be better to remove @DynaModule here and add install(new ReportModule()) in OnPremisesIdeApiModule.
@skabashnyuk Am I right?

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java (hunk 2 above)
Comment (skabashnyuk) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: very bad.
-1 for this class. It's too open.

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/UtilService.java (hunk 3 above)
Comment (skabashnyuk) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/UtilService.java]: Ok. Now I understand what is doing this method. Can you explain why? what for? What functionality it's covered?

File: wsmaster/codenvy-hosted-report/src/main/java/com/codenvy/report/ReportModule.java (hunk 16 above)
Comment (skabashnyuk) [inline wsmaster/codenvy-hosted-report/src/main/java/com/codenvy/report/ReportModule.java]: +1

File: wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java (hunk 10 above)
Comment (dmytro-ndp) [inline wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java]: That is the reminder: ndp is my nick, (codenvy/600) - number of issue.

File: wsmaster/codenvy-hosted-report-shared/pom.xml (hunk 12 above)
Comment (dmytro-ndp) [inline wsmaster/codenvy-hosted-report-shared/pom.xml]: So, how can I avoid it?

File: wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java (hunk 10 above)
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java]: 1. Dmytro Nochevnov is more cleaner for other developers
2. codenvy/600 issue says nothing about this TODO

I'd rewrite it as `//TODO Replace it with UserManager#getTotalCount when codenvy->jpa-integration branch will be merged to master`

File: wsmaster/codenvy-hosted-report-shared/src/main/java/com/codenvy/report/shared/dto/Ip.java (hunk 13 above)
Comment (dmytro-ndp) [inline wsmaster/codenvy-hosted-report-shared/src/main/java/com/codenvy/report/shared/dto/Ip.java]: It's just a value object which is in role of DTO now. So, I would prefer "Ip" name.

File: wsmaster/codenvy-hosted-report/pom.xml (hunk 14 above)
Comment (dmytro-ndp) [inline wsmaster/codenvy-hosted-report/pom.xml]: Do you think we could avoid it?

Comment (sleshchenko) [inline wsmaster/codenvy-hosted-report/pom.xml]: I guess yes

File: wsmaster/codenvy-hosted-report/src/main/java/com/codenvy/report/ReportModule.java (hunk 15 above)
Comment (dmytro-ndp) [inline wsmaster/codenvy-hosted-report/src/main/java/com/codenvy/report/ReportModule.java]: Thank you for advise.

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java [hunk 17]
```diff
@@ -47,18 +51,29 @@ public ReportService() {
...
-    public Response getReportParameters(@PathParam("report_type") String reportType) {
+    public Response getReportParameters(@PathParam("report_type") String reportTypeStr) {
         try {
-            if (reportType == null || ReportType.valueOf(reportType.toUpperCase()) == null) {
-                throw new RuntimeException("Report type is unknown.");
-            }
+            Objects.requireNonNull(reportTypeStr, "Report type is unknown.");
 
-            ReportParameters parameters = ReportType.valueOf(reportType.toUpperCase()).getParameters();
+            ReportType reportType = ReportType.valueOf(reportTypeStr.toUpperCase());
```
Comment (sleshchenko) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: IllegalArgumentException can be thrown here.

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java (hunk 7 above)
Comment (sleshchenko) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: When reportType can equal to null?

File: wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java (hunk 9 above)
Comment (dmytro-ndp) [inline wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java]: licenseText and licenseContent hardly ever helps us to defer the values for sure.

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java (hunk 2 above)
Comment (dmytro-ndp) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: What is bad? 
Report parameters isn't private info. 
This is Update Server, and we have a few public methods as well there.

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/UtilService.java (hunk 3 above)
Comment (dmytro-ndp) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/UtilService.java]: It is used to identify anonymous user in "Weekly Codenvy on-prem" report and for statistics.

File: wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/LicenseService.java (hunk 4 above)
Comment (dmytro-ndp) [inline wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/LicenseService.java]: Thank you for advice!

File: wsmaster/codenvy-hosted-report-shared/pom.xml (hunk 12 above)
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-report-shared/pom.xml]: Please explain dependency between modules which use these DTOs and mb then I'll can help you

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java (hunk 7 above)
Comment (dmytro-ndp) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: For @PathParam - never, as you mentioned earlier.

File: wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java (hunk 10 above)
Comment (dmytro-ndp) [inline wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java]: okay, thank you for ready to use variant :-)

File: wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java (hunk 11 above)
Comment (dmytro-ndp) [inline wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java]: What is wrong with formatting there?

File: wsmaster/codenvy-hosted-report/pom.xml (hunk 14 above)
Comment (dmytro-ndp) [inline wsmaster/codenvy-hosted-report/pom.xml]: Thank you, I will try.

File: wsmaster/codenvy-hosted-report/src/main/java/com/codenvy/report/ReportModule.java (hunk 16 above)
Comment (dmytro-ndp) [inline wsmaster/codenvy-hosted-report/src/main/java/com/codenvy/report/ReportModule.java]: Thank you for advice.

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java (hunk 17 above)
Comment (dmytro-ndp) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: I don't think so in case of usage @PathParam for getting **reportTypeStr** value.

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java (hunk 7 above)
Comment (sleshchenko) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: Even when reportTypeStr can be nullable reportType can't be nullable because Enum#valueOf method throws NPE in canse when enum name paremeter is nullable

File: wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java (hunk 11 above)
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java]: https://github.com/eclipse/che/blob/master/wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java#L126
```
  @parameterName
      parameter description here
```

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java (hunk 17 above)
Comment (sleshchenko) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: Please take a look java.lang.Enum#valueOf

Comment (dmytro-ndp) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: I will add the code to handle this exception. Thank you!

File: wsmaster/codenvy-hosted-report-shared/pom.xml (hunk 12 above)
Comment (dmytro-ndp) [inline wsmaster/codenvy-hosted-report-shared/pom.xml]: Module https://github.com/codenvy/codenvy/blob/d7d15310508c8e11e3ab3205710337e7bb395018/wsmaster/codenvy-hosted-report/pom.xml is using Ip.class to get IP value, and module https://github.com/codenvy/codenvy/blob/d7d15310508c8e11e3ab3205710337e7bb395018/cdec/update-server-packaging-war/pom.xml is using Ip class to send IP value.

File: wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java (hunk 11 above)
Comment (dmytro-ndp) [inline wsmaster/codenvy-hosted-api-license/src/main/java/com/codenvy/api/license/server/license/CodenvyLicenseManager.java]: I see. Thank you.

**Comment by dmytro-ndp** [comment]
@skabashnyuk, @sleshchenko: I fixed code according to your notices and advice as much as possible.

**Comment by dmytro-ndp** [comment]
@skabashnyuk, @sleshchenko: I fixed code according to your notices and advice.

**Comment by dmytro-ndp** [comment]
@tolusha: please, review this request.

File: wsmaster/codenvy-hosted-report-shared/pom.xml (hunk 12 above)
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-report-shared/pom.xml]: Ok for me

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java [hunk 18]
```diff
@@ -47,18 +53,32 @@ public ReportService() {
...
 
+    /** is package private for testing propose **/
+    ReportParameters getParameters(ReportType reportType) {
+        String parameterPrefix = reportType.name().toLowerCase();
+        ReportParameters parameters = new ReportParameters(
+            InjectorBootstrap.INJECTOR.getInstance(Key.get(String.class, Names.named(parameterPrefix + ".title"))),
+            InjectorBootstrap.INJECTOR.getInstance(Key.get(String.class, Names.named(parameterPrefix + ".sender"))),
+            InjectorBootstrap.INJECTOR.getInstance(Key.get(String.class, Names.named(parameterPrefix + ".receiver")))
+        );
+        return parameters;
```
Comment (sleshchenko) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]:
```
return new ReportParameters(
            InjectorBootstrap.INJECTOR.getInstance(Key.get(String.class, Names.named(parameterPrefix + ".title"))),
            InjectorBootstrap.INJECTOR.getInstance(Key.get(String.class, Names.named(parameterPrefix + ".sender"))),
            InjectorBootstrap.INJECTOR.getInstance(Key.get(String.class, Names.named(parameterPrefix + ".receiver")))
        );
```
?

Comment (dmytro-ndp) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: It is a kind of unified properties.

File: wsmaster/codenvy-hosted-report/src/main/java/com/codenvy/report/ReportSender.java [hunk 19]
```diff
@@ -0,0 +1,141 @@
...
+/**
+ * Sends reports:
+ * <li>to Codenvy with number of users.</li>
+ *
+ * @author Anatoliy Bazko
+ * @author Dmytro Nochevnov
+ */
+@Singleton
+public class ReportSender {
+    private static final Logger LOG = Logger.getLogger(ReportSender.class.getSimpleName());
```
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-report/src/main/java/com/codenvy/report/ReportSender.java]: It would be better to use slf4j logger here

**Comment by sleshchenko** [comment]
other OK for me

**Comment by dmytro-ndp** [comment]
@sleshchenko: please, review the fix.

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java (hunk 2 above)
Comment (gazarenkov) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: -1 as well, reports definitely contain sensitive info

**Comment by sleshchenko** [comment]
OK for me

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java (hunk 18 above)
Comment (sleshchenko) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: I mean that it would be better to inlune this code because you don't need `parameters` variable

File: cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java (hunk 2 above)
Comment (dmytro-ndp) [inline cdec/update-server-packaging-war/src/main/java/com/codenvy/im/update/ReportService.java]: ReportService returns next info:
`{"sender":"no-replay@codenvy.com",
"title":"Weekly On-Prem report",
"receiver":"data@codenvy.com"}`
Did you mean this info is a sensitive?
